tdmpc2/tdmpc2.py

```diff
@@ -218,48 +238,53 @@ def _td_target(self, next_z, reward, task):
 	def update(self, buffer):
 		"""
 		Main update function. Corresponds to one iteration of model learning.
-		
+
 		Args:
 			buffer (common.buffer.Buffer): Replay buffer.
-		
+
 		Returns:
 			dict: Dictionary of training statistics.
 		"""
 		obs, action, reward, task = buffer.sample()
-	
+		kwargs = {}
+		if task is not None:
+			kwargs["task"] = task
+		return self._update(obs, action, reward, **kwargs)
+
+	def _update(self, obs, action, reward, task=None):
 		# Compute targets
 		with torch.no_grad():
 			next_z = self.model.encode(obs[1:], task)
 			td_targets = self._td_target(next_z, reward, task)
 
 		# Prepare for update
-		self.optim.zero_grad(set_to_none=True)
+		self.optim.zero_grad()
 		self.model.train()
 
 		# Latent rollout
 		zs = torch.empty(self.cfg.horizon+1, self.cfg.batch_size, self.cfg.latent_dim, device=self.device)
 		z = self.model.encode(obs[0], task)
 		zs[0] = z
 		consistency_loss = 0
-		for t in range(self.cfg.horizon):
-			z = self.model.next(z, action[t], task)
-			consistency_loss += F.mse_loss(z, next_z[t]) * self.cfg.rho**t
+		for t, (_action, _next_z) in enumerate(zip(action.unbind(0), next_z.unbind(0))):
+			z = self.model.next(z, _action, task)
+			consistency_loss += F.mse_loss(z, _next_z) * self.cfg.rho**t
```

Comment:
here and below: I prefer unbind rather than indexing, it's usually faster (bc you access the original tensor just once to slice it)
but it should work with compile and cudagraphs without problems I think